### ISSUE TITLE
Bug with residual resources in layerStore after remove an image

### ISSUE THREAD

**Issue by zvier** [containers/storage#664]
Bug with residual resources in layerStore after remove an image may lead memory leak and affect other image build export failed with error messages as flows:

> ...Error reading blob sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef: error locating item named "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef" for image with ID "7f32b3d4368beaf08fc934a1e5fb24df607d31ec0b612d65f5807527363a3af9": file does not exist 

When delete an image, related layers in layerStore only be deleted in layers, idindex, byid, byname and bymount. But bycompressedsum and byuncompressedsum still maintain the relation for these layers utill the layerStore reload.

```go
// github.com/containers/storage/layers.go
type layerStore struct {
    ...
	layers            []*Layer
	idindex           *truncindex.TruncIndex
	byid              map[string]*Layer
	byname            map[string]*Layer
	bymount           map[string]*Layer
	bycompressedsum   map[digest.Digest][]string
	byuncompressedsum map[digest.Digest][]string
    ...
}
```

This may lead memory leak. And more seriously, these residual resources may cause image export failed in some special cases. My Dockerfile is:
```Dockerfile
FROM busybox:latest AS foo
ARG USED_ARG="used_value"
RUN echo ${USED_ARG} > /root/value
RUN echo ${USED_ARG}
COPY data /src/data

FROM foo
ENV http_proxy=xxxproxy
EXPOSE 80
```

The special in this Dockerfile is the second stage not produce any actual data content, so the diffid for this layer will always be a fixed value(In my environment is 5f70bf18a...). Therefore, in layerStore.byuncompressedsum will store the relation with the same key, for example, my layerStore.byuncompressedsum is ["sha256:5f70bf18..."]->["layerID_1"] at this time.

Then, we remove the image and layers about this image will be also removed in layerStore, except bycompressedsum and byuncompressedsum. For example, the layerStore.byuncompressedsum is still ["sha256:5f70bf18..."]->["layerID_1"] at this time.

The problem is that we may build and export an image use above Dockerfile again immediately. After the image build complete, layerStore.byuncompressedsum's value is ["sha256:5f70bf18..."]->["layerID_1", "layerID_2"] because it use append operation directly when newvalue is not empty.
```go
// github.com/containers/storage/layers.go
func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error) {
   ...
   if newvalue != ""  {
       (*m)[newvalue] = append((*m)[newvalue], id)
   }
   ...
}
```

And when export the image in the flow of storageImageSource.getBlobAndLayerID(), the layers of the image will not be found in layerStore and len(layers) == 0. Then use digest(sha256:5f70bf18...) as a key to retrieves the images's data will also failed which cause export image failed finally.
```go
// github.com/containers/image/blob/master/storage/storage_image.go
func (s *storageImageSource) getBlobAndLayerID(info types.BlobInfo) (rc io.ReadCloser, n int64, layerID string, err error) {
    ...
    layers, _ := s.imageRef.transport.store.LayersByUncompressedDigest(info.Digest)
    // If it's not a layer, then it must be a data item.
    if len(layers) == 0 {
	b, err := s.imageRef.transport.store.ImageBigData(s.image.ID, info.Digest.String())
	if err != nil {
		return nil, -1, "", err
	}
       ...
   }
   ...
}
```

Because when find layers with info.Digest, layerStore.byuncompressedsum's value is ["sha256:5f70bf18..."]->["layerID_1", "layerID_2"] at this time, but "layerID_1" is not exist in layerStore now, so will return (nil, ErrLayerUnknown) in this function call.
```go
// github.com/containers/storage/layers.go
func (r *layerStore) layersByDigestMap(m map[digest.Digest][]string, d digest.Digest) ([]Layer, error) {
	var layers []Layer
	for _, layerID := range m[d] {
		layer, ok := r.lookup(layerID)
		if !ok {
			return nil, ErrLayerUnknown
		}
		layers = append(layers, *copyLayer(layer))
	}
	return layers, nil
}
```

**Comment by zvier** [comment]
There have two ways which has test in my environment to solve this problem and I priored the first:

1. Delete the residual relation in layerStore.bycompressedsum and layerStore.byuncompressedsum when delete the layer.Advantage is operate in memory directly. Disadvantage is need to add more code compared with method 2.

```go
// github.com/containers/storage/layers.go
func (r *layerStore) deleteInternal(id string) error {
   ...
   if err == nil {
        ...
	    r.idindex.Delete(id)
		mountLabel := layer.MountLabel
		if layer.MountPoint != "" {
			delete(r.bymount, layer.MountPoint)
		}
        r.deleteInDigestMap(id)
        ...
}
```
```go
// github.com/containers/storage/layers.go
func (r *layerStore) deleteInDigestMap(id string) {
        for digest, layers := range r.bycompressedsum {
                for i, layerID := range layers {
                        if layerID == id {
                                layers = append(layers[:i], layers[i+1:]...)
                                r.bycompressedsum[digest] = layers
                                break
                        }
                }
        }
        for digest, layers := range r.byuncompressedsum {
                for i, layerID := range layers {
                        if layerID == id {
                                layers = append(layers[:i], layers[i+1:]...)
                                r.byuncompressedsum[digest] = layers
                                break
                        }
                }
        }
```

2. Reload the layerStore by call layerStore.Load() after delete an layer. Advantage is simple. Disadvantage is too heavy, and before Delete() is already call Load(), it seem strange call this method again.
```go
// github.com/containers/storage/layers.go
func (r *layerStore) Delete(id string) error {
        ...
        if r.Save() != nil {
                return err
        }
        return r.Load()
}
```

Any ideas or suggestions?

**Comment by zvier** [comment]
Fixed by #665
